importer/src/path_canonicalizer.rs

```diff
@@ -351,21 +354,21 @@ fn canonicalize_inside_type_scheme(
     );
 }
 
-fn canonicalize_inside_type(
-    ty: &mut Type<Expression>,
+fn canonicalize_inside_type<E>(
+    ty: &mut Type<E>,
     type_vars: &HashSet<&String>,
     path: &AbsoluteSymbolPath,
     paths: &'_ PathMap,
-) {
+) where
+    Type<E>: ExpressionVisitable<Expression>,
+{
```

Comment:
My understanding is that these are required because of the following implementation:

```rust
impl<Expr: ExpressionVisitable<Expr>, C: Children<Expr>> ExpressionVisitable<Expr> for C
```

`Type` implements `Children<Type>`, so if we set `C` and `Expr` to `Type`, we have a recursive trait dependency: `Type` implements `ExpressionVisitable<Type>` if it implements `ExpressionVisitable<Type>`.

This is fine as long as we are not calling methods from `ExpressionVisitable<Expr>` with `Expr` potentially being `Type`.

Below, when calling `pre_visit_expressions_mut`, even if we pass a closure which takes an `Expression` as input, the compiler still tries to set `Expr` to `Type`, which triggers the recursive dependency. Note that this still happens even though we pass a closure which takes `Expression` as input (and therefore excludes `ExpressionVisitable<Type>` which would require a closure taking a `Type`), so it seems like the compiler is not using this knowledge.

Now the confusing part: adding the where clause somehow tells the compiler not to try to apply `ExpressionVisitable<Type>`, even though it is implemented!